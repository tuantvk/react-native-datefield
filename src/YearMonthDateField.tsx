import React, { createRef } from 'react';
import { View, Keyboard, TextInput } from 'react-native';
import {
  int,
  isValidDate,
  formatYearDigits,
  getOnlyNumber,
  getDateDefault,
  daysInMonth,
  dateInRange,
} from './utils';
import Input from './Input';
import type { DateFieldProps } from './types';
import styles from './styles';

type State = {
  date: string;
  month: string;
  year: string;
};

class YearMonthDateField extends React.Component<DateFieldProps, State> {
  static defaultProps = {
    labelDate: 'Date',
    labelMonth: 'Month',
    labelYear: 'Year',
    editable: true,
  };

  state = { ...getDateDefault(this.props.defaultValue) };

  refDate = createRef<TextInput>();
  refMonth = createRef<TextInput>();
  refYear = createRef<TextInput>();

  componentDidMount() {
    if (this.props.autoFocus) {
      this.refYear.current?.focus();
    }
  }

  UNSAFE_componentWillReceiveProps(nextProps: DateFieldProps) {
    if (
      JSON.stringify(getDateDefault(nextProps.value)) !==
      JSON.stringify(getDateDefault(this.props.value))
    ) {
      const { date, month, year } = getDateDefault(nextProps.value);
      const nextState = {
        date: date ? date.padStart(2, '0') : '',
        month: month ? month.padStart(2, '0') : '',
        year,
      };
      this.setState(nextState);
    }
  }

  onChangeDate = (value: string) => {
    const date = getOnlyNumber(int(value) > 31 ? '31' : value);
    this.setState({ date }, () => {
      if (date.length === 2) {
        Keyboard.dismiss();
      }
    });
  };

  onChangeMonth = (value: string) => {
    const month = getOnlyNumber(int(value) > 12 ? '12' : value);
    this.setState({
      month,
      date: daysInMonth(this.state),
    });
    if (month.length === 2) {
      this.refDate.current?.focus();
    }
  };

  onChangeYear = (value: string) => {
    const year = getOnlyNumber(value);
    this.setState({ year });
    if (year.length === 4) {
      this.refMonth.current?.focus();
    }
  };

  onBlur = () => {
    const { maximumDate, minimumDate, handleErrors, onSubmit } = this.props;
    const current = { ...this.state };
    if (int(current.date) === 0) {
      current.date = '01';
    }
    if (current.date.length === 1) {
      current.date = current.date.padStart(2, '0');
    }
    if (int(current.month) === 0) {
      current.month = '01';
    }
    if (current.month.length === 1) {
      current.month = current.month.padStart(2, '0');
    }
    if (int(current.year) === 0) {
      current.year = `${new Date().getFullYear()}`;
    }
    if (current.year.length > 1 && current.year.length < 4) {
      current.year = `${formatYearDigits(int(current.year))}`;
    }
    const value = new Date(
      int(current.year),
      int(current.month) - 1, // new Date(year, monthIndex, day)
      int(current.date)
    );
    if (current.date) {
      if (
        (minimumDate || maximumDate) &&
        !dateInRange(value, minimumDate, maximumDate)
      ) {
        handleErrors && handleErrors();
        this.setState({ date: '', month: '', year: '' });
      } else {
        if (isValidDate(value)) {
          onSubmit && onSubmit(value);
        }
        this.setState({ ...current });
      }
    }
  };

  render() {
    const { date, month, year } = this.state;
    const {
      testID,
      containerStyle,
      styleInput,
      styleInputYear,
      labelDate,
      labelMonth,
      labelYear,
      editable,
      placeholderTextColor,
    } = this.props;

    return (
      <View {...{ testID }} style={[styles.container, containerStyle]}>
        <Input
          ref={this.refYear}
          value={year}
          maxLength={4}
          placeholder={labelYear}
          style={styleInput}
          onChangeText={this.onChangeYear}
          onSubmitEditing={() => this.refMonth.current?.focus()}
          onBlur={this.onBlur}
          {...{ editable, placeholderTextColor }}
        />
        <Input
          ref={this.refMonth}
          value={month}
          placeholder={labelMonth}
          style={styleInput}
          onChangeText={this.onChangeMonth}
          onSubmitEditing={() => this.refDate.current?.focus()}
          onBlur={this.onBlur}
          {...{ editable, placeholderTextColor }}
        />
        <Input
          ref={this.refDate}
          value={date}
          returnKeyType="done"
          placeholder={labelDate}
          style={[styleInput, styleInputYear]}
          onChangeText={this.onChangeDate}
          onSubmitEditing={() => Keyboard.dismiss()}
          onBlur={this.onBlur}
          {...{ editable, placeholderTextColor }}
        />
      </View>
    );
  }
}

export default YearMonthDateField;
