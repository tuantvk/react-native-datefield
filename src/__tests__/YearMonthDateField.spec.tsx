import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import YearMonthDateField from '../YearMonthDateField';

describe('YearMonthDateField component', () => {
  it('should match snapshot', () => {
    const component = shallow(<YearMonthDateField />);
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  describe('Props', () => {
    it('testID', () => {
      const component = shallow(<YearMonthDateField testID="date_input" />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('containerStyle', () => {
      const component = shallow(
        <YearMonthDateField containerStyle={{ marginVertical: 20 }} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('styleInput', () => {
      const component = shallow(
        <YearMonthDateField styleInput={{ fontSize: 15 }} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelDate', () => {
      const component = shallow(<YearMonthDateField labelDate="Input date" />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelMonth', () => {
      const component = shallow(
        <YearMonthDateField labelMonth="Input month" />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('labelYear', () => {
      const component = shallow(<YearMonthDateField labelYear="Input year" />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('defaultValue', () => {
      const component = shallow(
        <YearMonthDateField defaultValue={new Date('2021-03-25')} />
      );
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('onSubmit', () => {
      const component = shallow(<YearMonthDateField onSubmit={() => {}} />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it('editable', () => {
      const component = shallow(<YearMonthDateField editable={false} />);
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
