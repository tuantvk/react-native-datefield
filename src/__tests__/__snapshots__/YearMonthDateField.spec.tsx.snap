// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearMonthDateField component Props containerStyle 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      Object {
        "marginVertical": 20,
      },
    ]
  }
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    value=""
  />
</View>
`;

exports[`YearMonthDateField component Props defaultValue 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value="2021"
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    value="3"
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    value="25"
  />
</View>
`;

exports[`YearMonthDateField component Props editable 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <ForwardRef
    editable={false}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
  <ForwardRef
    editable={false}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    value=""
  />
  <ForwardRef
    editable={false}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    value=""
  />
</View>
`;

exports[`YearMonthDateField component Props labelDate 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input date"
    returnKeyType="next"
    value=""
  />
</View>
`;

exports[`YearMonthDateField component Props labelMonth 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input month"
    returnKeyType="next"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    value=""
  />
</View>
`;

exports[`YearMonthDateField component Props labelYear 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input year"
    returnKeyType="done"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    value=""
  />
</View>
`;

exports[`YearMonthDateField component Props onSubmit 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    value=""
  />
</View>
`;

exports[`YearMonthDateField component Props styleInput 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    style={
      Object {
        "fontSize": 15,
      }
    }
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    style={
      Object {
        "fontSize": 15,
      }
    }
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    style={
      Object {
        "fontSize": 15,
      }
    }
    value=""
  />
</View>
`;

exports[`YearMonthDateField component Props testID 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
  testID="date_input"
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    value=""
  />
</View>
`;

exports[`YearMonthDateField component should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      undefined,
    ]
  }
>
  <ForwardRef
    editable={true}
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Year"
    returnKeyType="done"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Month"
    returnKeyType="next"
    value=""
  />
  <ForwardRef
    editable={true}
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Date"
    returnKeyType="next"
    value=""
  />
</View>
`;
